### PR TITLE
[SDPT-533] write files to nfs external mount

### PR DESCRIPTION
**Pull request by chennv4**
**Feature:**
Mount additional NFS mount and write files to external mount

**Solution:**
Detailed steps mentioned in README

**Results:**
Tested on http://valkyrie.em.sdp.hop.lab.emc.com/dashboard

# ls -l /mnt/examples-sample/2020-07-15--07
total 2706
-rw-r--r-- 1 root root  847 Jul 15 07:00 part-0-109
-rw-r--r-- 1 root root 1330 Jul 15 07:00 part-0-110
-rw-r--r-- 1 root root 1088 Jul 15 07:00 part-0-111
-rw-r--r-- 1 root root 1208 Jul 15 07:00 part-0-112
-rw-r--r-- 1 root root 1087 Jul 15 07:00 part-0-113
-rw-r--r-- 1 root root 1331 Jul 15 07:00 part-0-114
-rw-r--r-- 1 root root 1208 Jul 15 07:01 part-0-115

### CODE REVIEW

File: charts/flink-tools/values.yaml [hunk 1]
```diff
@@ -31,12 +31,14 @@ jobManager:
...
   memory: "1024M"
   replicas: 1
+
 taskManager:
   replicas: 1
   heap: "3221225472"      # 3Gi
   memory: "4294967296"    # 4Gi
   cpu: "250m"
   numberOfTaskSlots: 1
+
```
Comment (claudiofahey) [inline charts/flink-tools/values.yaml:44]: Add `enableExtraVolume: false` to this file. It will become the default value. Then you can remove this from all other yaml files.

File: README.md [hunk 2]
```diff
@@ -14,6 +14,7 @@ Pravega Flink Tools is a collection of Apache Flink applications for working wit
 It provides the following Flink jobs:
 
 - **stream-to-file**: Continuously copy a Pravega stream to files on S3, HDFS, or any other Flink-supported file system
+- **stream-to-nfs**: Continuously copy a Pravega stream to files on external NFS mount
```
Comment (claudiofahey) [inline README.md]: Remove this line.

File: README.md [hunk 3]
```diff
@@ -14,6 +14,7 @@ Pravega Flink Tools is a collection of Apache Flink applications for working wit
 It provides the following Flink jobs:
 
 - **stream-to-file**: Continuously copy a Pravega stream to files on S3, HDFS, or any other Flink-supported file system
```
Comment (claudiofahey) [inline README.md]: Change to: Continuously copy a Pravega stream to files on S3, HDFS, NFS, or any other Flink-supported file system

File: README.md [hunk 4]
```diff
@@ -67,7 +68,50 @@ For binary events, you will need to customize the Flink job with the appropriate
 Flink offers many options for customizing the behavior when writing files.
 Refer to [Steaming File Sink](https://ci.apache.org/projects/flink/flink-docs-release-1.10/dev/connectors/streamfile_sink.html)
 for details.
+## Configuring a seperate NFS mount to write files to NFS
```
Comment (claudiofahey) [inline README.md]: Spelled "separate".

File: README.md [hunk 5]
```diff
@@ -67,7 +68,50 @@ For binary events, you will need to customize the Flink job with the appropriate
...
+apiVersion: v1
+metadata:
+  name: flink-tools-nfs
+spec:
+  accessModes:
+    - ReadWriteMany
+  resources:
+    requests:
+      storage: 10Gi
+  storageClassName: nfs
```
Comment (claudiofahey) [inline README.md]: This will use the same Isilon used by Pravega LTS. The requirement is:
> Create a persistent volume that points to an arbitrary NFS mount and path. This must be different from any NFS mount used for Pravega LTS.
I believe you will need to create a new storage class that uses the NFS client provisioner that points to a different NFS server.
Also, don't worry about creating Helm templates for the PVC. Just create a sample yaml file. We'll let the user be responsible for that prior to running Helm.

File: charts/flink-tools/templates/FlinkCluster.yaml [hunk 6]
```diff
@@ -5,6 +5,20 @@
 # You may obtain a copy of the License at
 #
 #     http://www.apache.org/licenses/LICENSE-2.0
+{{- if .Values.enableExtraVolume }}
```
Comment (claudiofahey) [inline charts/flink-tools/templates/FlinkCluster.yaml]: As mentioned earlier, do not create the PVC in Helm.

File: charts/flink-tools/templates/FlinkCluster.yaml [hunk 7]
```diff
@@ -22,6 +36,12 @@ spec:
         requests:
           storage: {{ .Values.localStorage.size | quote }}
       storageClassName: standard
+  {{- if .Values.enableExtraVolume }}
```
Comment (claudiofahey) [inline charts/flink-tools/templates/FlinkCluster.yaml]: This is good.

File: values/job-defaults/stream-to-nfs-job.yaml [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+enableExtraVolume: true
+appParameters:
+  # Files will be created with this interval.
+  checkpointIntervalMs: "60000"
+mainClass: io.pravega.flinktools.StreamToFileJob
+jobManager:
+  volumeMounts:
+    - mountPath: /mnt/examples-sample
```
Comment (claudiofahey) [inline values/job-defaults/stream-to-nfs-job.yaml]: This file should contain job defaults, not examples. Remove the jobManager and taskManager sections.

File: values/samples/sample1-stream-to-nfs-job.yaml [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+  checkpointTimeoutMs: "3600000"  # 1 hr
+  input-stream: "sample1"
+  #input-startAtTail: "false"
+  #input-startStreamCut: "UNBOUNDED"
+  output: "/mnt/examples-sample"
+  #parallelism: "1"
+imageRef:
+  name: "1.10.0-2.12-hadoop2.8.3"
+clusterConfiguration:
+  taskmanager.memory.jvm-metaspace.size: "536870912"  # 512 MiB
```
Comment (claudiofahey) [inline values/samples/sample1-stream-to-nfs-job.yaml:20]: Add the volumeMounts to this file instead.

**Review by claudiofahey — CHANGES_REQUESTED** [review]

File: README.md (hunk 2 above)
Comment (chennv4) [reply README.md]: removed

File: README.md (hunk 3 above)
Comment (chennv4) [reply README.md]: modified

File: README.md (hunk 4 above)
Comment (chennv4) [reply README.md]: changed

File: charts/flink-tools/values.yaml (hunk 1 above)
Comment (chennv4) [reply charts/flink-tools/values.yaml:44]: added

File: README.md (hunk 5 above)
Comment (chennv4) [reply README.md]: Storage Class is not required. We have already nfs storage class on our cluster. We can use the same.
chenna@chenna:~/trail/sdpt-533/flink-tools(feature-sdpt-533-write-files-to-nfs)$ k get sc
NAME                 PROVISIONER                            AGE
nfs                  cluster.local/nfs-client-provisioner   5d
standard (default)   kubernetes.io/vsphere-volume           5d5h

File: charts/flink-tools/templates/FlinkCluster.yaml (hunk 6 above)
Comment (chennv4) [reply charts/flink-tools/templates/FlinkCluster.yaml]: created separate sample files for PV and PVC

File: values/job-defaults/stream-to-nfs-job.yaml (hunk 8 above)
Comment (chennv4) [reply values/job-defaults/stream-to-nfs-job.yaml]: removed from job-defaults and added in samples.

File: values/samples/sample1-stream-to-nfs-job.yaml (hunk 9 above)
Comment (chennv4) [reply values/samples/sample1-stream-to-nfs-job.yaml:20]: added

**Comment by chennv4** [comment]
Modified and tested as per review comments.

File: README.md [hunk 10]
```diff
@@ -67,7 +67,51 @@ For binary events, you will need to customize the Flink job with the appropriate
...
+  nfs:
+    path: <MOUNT_PATH>
+    server: <NFS_SERVER>
+
+# Execute these from ~/flink-tools
+k create -f charts/external-nfs/template/external-nfs-pv.yaml
+
+k create -f charts/external-nfs/template/external-nfs-pvc.yaml -n <NAMESPACE>
+
+# deploy flink job for JFS
```
Comment (claudiofahey) [inline README.md]: Fix typo "JFS" to "NFS".

File: charts/external-nfs/template/external-nfs-pv.yaml [hunk 11]
```diff
@@ -0,0 +1,18 @@
+apiVersion: v1
```
Comment (claudiofahey) [inline charts/external-nfs/template/external-nfs-pv.yaml]: This is not a valid Helm chart. Relocate this to /nfs-samples.

File: charts/flink-tools/templates/FlinkCluster.yaml [hunk 12]
```diff
@@ -22,6 +22,12 @@ spec:
         requests:
           storage: {{ .Values.localStorage.size | quote }}
       storageClassName: standard
+{{- if .Values.enableExtraVolume }}
```
Comment (claudiofahey) [inline charts/flink-tools/templates/FlinkCluster.yaml]: This should be changed to support multiple NFS volumes. Change this to:
```
  volumes:
{{ toYaml .Values.volumes | indent 4 }}
```

Then in /charts/flink-tools/values.yaml, specify an empty default:
```
volumes: {}
```

File: charts/flink-tools/values.yaml [hunk 13]
```diff
@@ -22,6 +22,8 @@ localStorage:
   replicas: 2
   size: "5G"
 #mainClass: Name of main class. REQUIRED.
+enableExtraVolume: false
```
Comment (claudiofahey) [inline charts/flink-tools/values.yaml]: This flag will be unneeded with above change.

File: charts/flink-tools/values.yaml [hunk 14]
```diff
@@ -31,13 +33,16 @@ jobManager:
...
+
 taskManager:
   replicas: 1
   heap: "3221225472"      # 3Gi
   memory: "4294967296"    # 4Gi
   cpu: "250m"
   numberOfTaskSlots: 1
+
 clusterConfiguration:
+  restart-strategy.fixed-delay.delay: "30 s"
```
Comment (claudiofahey) [inline charts/flink-tools/values.yaml]: Remove this unrelated change.

File: scripts/jobs/stream-to-nfs-job.sh [hunk 15]
```diff
@@ -0,0 +1,29 @@
+#! /bin/bash
```
Comment (claudiofahey) [inline scripts/jobs/stream-to-nfs-job.sh]: Remove this script. The existing script stream-to-file.job.sh can be used.

File: values/job-defaults/stream-to-nfs-job.yaml [hunk 16]
```diff
@@ -0,0 +1,12 @@
+# Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
```
Comment (claudiofahey) [inline values/job-defaults/stream-to-nfs-job.yaml]: Remove this file.

File: values/job-defaults/stream-to-stream-job.yaml [hunk 17]
```diff
@@ -5,6 +5,7 @@
 # You may obtain a copy of the License at
 #
 #     http://www.apache.org/licenses/LICENSE-2.0
+enableExtraVolume: false
```
Comment (claudiofahey) [inline values/job-defaults/stream-to-stream-job.yaml]: Remove this flag.

**Review by claudiofahey — CHANGES_REQUESTED** [review]

File: README.md [hunk 18]
```diff
@@ -70,35 +70,23 @@ for details.
...
+# Edit values/nfs-samples/external-nfs-pv.yaml
 namespace: <NAMESPACE>
     name: flink-tools-nfs-pvc
   nfs:
     path: <MOUNT_PATH>
     server: <NFS_SERVER>
 
 # Execute these from ~/flink-tools
-k create -f charts/external-nfs/template/external-nfs-pv.yaml
+k create -f values/nfs-samples/external-nfs-pv.yaml
```
Comment (claudiofahey) [inline README.md]: Replace `k` with `kubectl` on this line and other lines. Most users will not have this alias.

File: README.md [hunk 19]
```diff
@@ -70,35 +70,23 @@ for details.
...
 
-k create -f charts/external-nfs/template/external-nfs-pvc.yaml -n <NAMESPACE>
+k create -f values/nfs-samples/external-nfs-pvc.yaml -n <NAMESPACE>
```
Comment (claudiofahey) [inline README.md]: Replace `<NAMESPACE>` with `${NAMESPACE}`. This allows one to set environment variables and copy/paste the command exactly as shown.

File: charts/flink-tools/templates/FlinkCluster.yaml [hunk 20]
```diff
@@ -22,12 +22,8 @@ spec:
         requests:
           storage: {{ .Values.localStorage.size | quote }}
       storageClassName: standard
-{{- if .Values.enableExtraVolume }}
   volumes:
-    - name: extra-volume
-      persistentVolumeClaim:
-        claimName: flink-tools-nfs-pvc
-{{- end }}
+  {{ toYaml .Values.volumes | indent 4 }}
```
Comment (claudiofahey) [inline charts/flink-tools/templates/FlinkCluster.yaml]: Check your spacing here. I believe that this needs to start at the first column, like line 33.

File: README.md [hunk 21]
```diff
@@ -70,35 +70,23 @@ for details.
 ## Configuring a external NFS mount to write files to external NFS mount.
```
Comment (claudiofahey) [inline README.md]: Replace line 70 with this text:

## Writing to an NFS volume

Use this procedure to configure the Flink Stream to File job to write to any Kubernetes Persistent Volume, such as a remote NFS volume.

File: README.md [hunk 22]
```diff
@@ -70,35 +70,23 @@ for details.
...
-k create -f charts/external-nfs/template/external-nfs-pvc.yaml -n <NAMESPACE>
+k create -f values/nfs-samples/external-nfs-pvc.yaml -n <NAMESPACE>
 
-# deploy flink job for JFS
-UNINSTALL=1 scripts/jobs/stream-to-nfs-job.sh values/samples/sample1-stream-to-nfs-job.yaml
+# deploy flink job for NFS
+UNINSTALL=1 scripts/jobs/stream-to-file-job.sh values/samples/sample1-stream-to-nfs-job.yaml
 ```
-2. Configure volumes spec in FlinkCluster.yaml
+2. Configure extra NFS moount in values/samples/sample1-stream-to-nfs-job.yaml
```
Comment (claudiofahey) [inline README.md]: Fix typo `moount` to `mount`.

File: values/job-defaults/stream-to-stream-job.yaml (hunk 17 above)
Comment (claudiofahey) [reply values/job-defaults/stream-to-stream-job.yaml]: This flag is still there.

**Review by claudiofahey — CHANGES_REQUESTED** [review]

File: README.md (hunk 10 above)
Comment (chennv4) [reply README.md]: fixed

File: charts/external-nfs/template/external-nfs-pv.yaml (hunk 11 above)
Comment (chennv4) [reply charts/external-nfs/template/external-nfs-pv.yaml]: moved to nfs-samples

File: charts/flink-tools/templates/FlinkCluster.yaml (hunk 12 above)
Comment (chennv4) [reply charts/flink-tools/templates/FlinkCluster.yaml]: changed accordingly

File: charts/flink-tools/values.yaml (hunk 13 above)
Comment (chennv4) [reply charts/flink-tools/values.yaml]: removed

File: charts/flink-tools/values.yaml (hunk 14 above)
Comment (chennv4) [reply charts/flink-tools/values.yaml]: removed

File: scripts/jobs/stream-to-nfs-job.sh (hunk 15 above)
Comment (chennv4) [reply scripts/jobs/stream-to-nfs-job.sh]: removed

File: values/job-defaults/stream-to-nfs-job.yaml (hunk 16 above)
Comment (chennv4) [reply values/job-defaults/stream-to-nfs-job.yaml]: removed

File: values/job-defaults/stream-to-stream-job.yaml (hunk 17 above)
Comment (chennv4) [reply values/job-defaults/stream-to-stream-job.yaml]: removed

File: README.md (hunk 18 above)
Comment (chennv4) [reply README.md]: changed

File: README.md (hunk 19 above)
Comment (chennv4) [reply README.md]: changed

File: charts/flink-tools/templates/FlinkCluster.yaml (hunk 20 above)
Comment (chennv4) [reply charts/flink-tools/templates/FlinkCluster.yaml]: Fixed

File: README.md (hunk 21 above)
Comment (chennv4) [reply README.md]: Changed

File: README.md (hunk 22 above)
Comment (chennv4) [reply README.md]: Fixed

**Review by claudiofahey — APPROVED** [review]
Looks good!
